Add global logging of HTTP operations

dataclouder opened the pull request:
Added a global log management utility (in the new util package)
that regulates logging in the whole library.

Logging is disabled by default and it can be enabled and directed to
either a file or the screen using public variables or environment
variables.
When enabled, all HTTP requests and responses are logged automatically.

Logging is documented in util/LOGGING.md

Changed in the library: 
* added calls to util processing functions into api.go.
* added util/logging.go

Implements Issue #68.
Updates Issue #67 and partially alleviates problems from Issue #38.

Signed-off-by: Giuseppe Maxia

auppunda commented on govcd/api.go:
```diff
@@ -51,6 +53,13 @@ func (c *Client) NewRequest(params map[string]string, method string, u url.URL,
 	}
 
+	payload := ""
+	if req.ContentLength > 0 {
```
Could you put a comment explaining these changes in api.go

auppunda commented on util/logging.go:
```diff
@@ -0,0 +1,205 @@
+	envLogOnScreen = "GOVCD_LOG_ON_SCREEN"
+
+	// Name of the environment variable that enables logging of passwords
+	envLogPasswords = "GOVCD_LOG_PASSWORDS"
```
Do you set GOVCD_LOG_PASSWORDS to true ?

dataclouder replied:
no. 
This is the name of the environment variable to change the default, which is
```LogPasswords bool = false```

lvirbalas commented on util/LOGGING.md:
````diff
@@ -0,0 +1,75 @@
+util.SetLog()
+```
+
+When enabled, the default output for logging is a file named `govcd-API-YYYY-MM-DD.HH.log`.
````
I was thinking about the default frequency of the logs generated. We want the defaults to be easy to use and not overload the user. It's now generating a log file each hour, but even if we reduce it to getting a new log each day it's still too noisy for a default.

Please change the default log filename to `go-vcloud-director.log` without date/time reference, but add a write-up that the YYYY-MM-DD, etc. are supported, so the advanced clients know that's available.

BTW, this will make it consistent with the Python vcd CLI default logging.

auppunda commented:
An idea that I would like the rest of the teams opinion on: Having an env variable to limit payloads. As in no XML payload greater will be dumped

lvirbalas commented:
@auppunda, if we limit how much of the payload we show, we risk not having enough information when we actually need it, hence it feels cleaner to implement #77 which would allow us to spit out the payload only when an error occurs.

dataclouder commented:
There is no risk in adding the `GOVCD_MAX_PAYLOAD_SIZE` variable. It's up to the user to use it or not. It doesn't cost much in terms of code (it's already done, in fact.)

hodgesrm commented:
@auppunda I think we should not add an extra variable to control length of payloads.  I would instead think of ways we can take variables away. (Back to an in important principle in Ousterhout's 'Philosophy of Software Design.')

dataclouder commented:
@lvirbalas, @auppunda, @hodgesrm 
I threw away the extra variables (by the time Ankith was outvoted I had already added a size limit for response text) and added comments where requested.
I have also added a check to detect whether a payload contains binary data instead of text (using the headers contents.)
The log file is now named `go-vcloud-director.log`.

hodgesrm commented on util/logging.go:
```diff
@@ -0,0 +1,242 @@
+	// Sends log to screen. If value is either "stderr" or "err"
+	// logging will go to os.Stderr. For any other value it will
+	// go to os.Stdout
+	LogOnScreen string = ""
```
It would be more flexible if we allowed clients to pass in their own Logger.  This would allow the SDK to integrate with whatever logging scheme the client uses.   If they don't we use the options to configure a default logger.

auppunda replied:
Are you suggesting for us to create a Logger interface as well, so that clients can add function implementations if they are not satisfied with the default?

auppunda replied:
Ya I think logging is one of those things where everybody has their own preferences, so allowing the client to choose how go-vcloud-director logs is probably a good thing

dataclouder replied:
@hodgesrm if we use a customized logger, we would have control only on what we do within the logging module. My purpose is also to alter the global behavior of `log.anyfunction` so that the output is the same for all the occurrence of logging within the library.
Would it be acceptable to pass instead an `io.writer`?
https://golang.org/pkg/log/#SetOutput

dataclouder replied:
On second thought, I should probably go the extra mile and use a private logger for our module. This will allow our logs to not mix with client logging. Users can change it by giving us their preferred logger.
Will be back soon with a cleaner implementation.

hodgesrm replied:
OK, sounds as if you are on the case.  I also thought about the io.writer approach and even wrote it in my comment and then erased it before pressing Comment. :)  

As far as config vs. environmental variables I notice Kubernetes API clients mostly use a single environmental variable that points to a config file ($KUBECONFIG). It would be worth checking out some other Golang API clients to see what they do.

dataclouder replied:
I have reworked the logging module to use a private logger, and to accept a custom logger from users.
I have also cleaned up the library from all the occurrences of `log.Printf`, changing them to `util.GovcdLogger.Printf`.

I have not, however, introduced a configuration file. I feel that configuration files should belong in the client, not the library. We use a configuration file in our test, which is, for this purpose, a client of the library.

The next iteration on testing (already in progress) will introduce a `log` section in the test configuration file.

An updated commit will come soon.
